iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -879,24 +932,35 @@ struct MakeDispatchWorkgroupsOp : public RewritePattern {
 
 /// Checks if the `producer` can be fused into the dispatch region of the
 /// `consumer`.
-static bool isProducerFusableWithConsumer(linalg::LinalgOp producer,
-                                          linalg::LinalgOp consumer) {
-  // Fuse if the dependence from producer to consumer is to the `outs` operand
-  // of the consumer and the producer is a parallel operation
-  llvm::DenseSet<Value> producerResults(producer.getOperation()->result_begin(),
-                                        producer.getOperation()->result_end());
-  auto isProducerResultValue = [&producerResults](OpOperand &operand) -> bool {
-    return producerResults.count(operand.get());
-  };
-  auto isElementWiseParallelOp = [](linalg::LinalgOp op) -> bool {
+static bool isProducerFusable(linalg::LinalgOp producer,
+                              linalg::LinalgOp consumer,
+                              OpOperand &consumerOperand) {
+  auto isElementWiseParallelOp = [](linalg::LinalgOp op) {
     return llvm::all_of(op.iterator_types(), [](Attribute attr) {
-      return attr.cast<StringAttr>().getValue() ==
-             toString(IteratorType::Parallel);
+      return linalg::isParallelIteratorType(attr);
     });
   };
-  return llvm::none_of(consumer.getInputOpOperands(), isProducerResultValue) &&
-         llvm::any_of(consumer.getOutputOpOperands(), isProducerResultValue) &&
-         isElementWiseParallelOp(producer);
+
+  if (consumer.isInputTensor(&consumerOperand)) {
+    // Make sure that we have an identity indexing map for the operand for now.
+    //
+    // Theoretically with tensor abstraction, we can pull in whatever operand
+    // subrange for fusion. But if the consumer's access patterns for the input
+    // and output are not the same, we must materialize buffers for holding the
+    // temporary result, because we can have multiple levels of distribution.
+    // Identity map makes sure that we can actually avoid a new intermediate
+    // buffer and directly use the one for the consumer.
+    auto map = consumer.getIndexingMap(consumerOperand.getOperandNumber());
+    if (!map.isIdentity()) return false;
+
+    // If the producer's result is used by the consumer as an input, fuse if
+    // this producer has no other users.
+    return consumerOperand.get().hasOneUse();
+  } else {
+    // If the producer's result is used by the consumer as an output
+    // initializer, fuse if the producer is a parallel operation.
+    return isElementWiseParallelOp(producer);
```

Comment:
The comment is fairly disjoint w.r.t what I read as the the implementation of `isElementWiseParallelOp`

Reply:
Moved `isElementWiseParallelOp` down here and modified the comment a bit.